lib/SILOptimizer/Mandatory/Differentiation.cpp

```diff
@@ -4418,23 +4516,95 @@ class AdjointEmitter final : public SILInstructionVisitor<AdjointEmitter> {
           getPullbackInfo().lookUpPullbackStructPredecessorField(bb);
       auto *predEnumVal =
           builder.createStructExtract(adjLoc, pbStructVal, predEnumField);
+
+      // Propagate adjoint values from active basic block arguments to
+      // predecessor terminator operands.
+      for (auto *bbArg : bb->getArguments()) {
+        if (!getActivityInfo().isActive(bbArg, getIndices()))
+          continue;
+        // Get predecessor terminator operands.
+        SmallVector<std::pair<SILBasicBlock *, SILValue>, 4> incomingValues;
+        bbArg->getSingleTerminatorOperands(incomingValues);
+        // Initialize adjoint value of predecessor terminator operands as
+        // adjoint value of current block arguments.
+        auto bbArgAdj = getAdjointValue(bb, bbArg);
+        for (auto pair : incomingValues) {
+          auto *predBB = std::get<0>(pair);
+          auto incomingValue = std::get<1>(pair);
+          initializeAdjointValue(predBB, incomingValue, bbArgAdj);
+        }
+      }
+
+      // 3. Build the adjoint successor cases for the `switch_enum` instruction.
+      //    The adjoint successors correspond to the predecessors of the current
+      //    block.
       SmallVector<std::pair<EnumElementDecl *, SILBasicBlock *>, 4>
-          predecessorCases;
+          adjointSuccessorCases;
       for (auto *predBB : bb->getPredecessorBlocks()) {
-        auto *adjointSuccBB = getAdjointBlock(predBB);
+        // Get the adjoint block and optional adjoint trampoline block of the
+        // predecessor block.
+        auto *adjointBB = getAdjointBlock(predBB);
+        auto *adjointTrampolineBB = getAdjointTrampolineBlock(predBB, bb);
+        SILBasicBlock *adjointSuccBB = nullptr;
+        // If the predecesssor block does not have a corresponding adjoint
+        // trampoline block, then the adjoint successor is the adjoint block.
+        if (!adjointTrampolineBB) {
+          adjointSuccBB = adjointBB;
+        }
+        // Otherwise, the adjoint successor is the adjoint trampoline block,
+        // which branches to the adjoint block and propagates adjoint values of
+        // active values.
+        else {
+          adjointSuccBB = adjointTrampolineBB;
+          adjointSuccBB = getAdjointTrampolineBlock(predBB, bb);
+          assert(adjointSuccBB && adjointSuccBB->getNumArguments() == 1);
+          SILBuilder adjointTrampolineBBBuilder(adjointSuccBB);
+          // Propagate pullback struct argument.
+          SmallVector<SILValue, 8> arguments;
+          arguments.push_back(adjointSuccBB->getArguments().front());
+          // Propagate adjoint values/buffers of active values/buffers.
+          for (unsigned i : indices(activeValues[predBB])) {
```

Comment:
Note: this adjoint value/buffer propagation logic should be hoisted out of this `if-else` block so that it executes regardless of whether there's a trampoline block.

Reply:
Furthermore, it's necessary to propagate more than just the "adjoint values of the active values in `predBB`".
Consider the following function:
```swift
@differentiable
@_silgen_name("cond_empty")
func cond_empty(_ x: Float) -> Float {
  if x > 0 {
    // Generates empty basic blocks between entry/exit.
  }
  return x
}
```

```
// cond_empty
sil hidden [differentiable source 0 wrt 0 jvp @AD__cond_empty__jvp_src_0_wrt_0 vjp @AD__cond_empty__vjp_src_0_wrt_0] @cond_empty : $@convention(thin) (Float) -> Float {
// %0                                             // users: %5, %20, %1
bb0(%0 : $Float):
  debug_value %0 : $Float, let, name "x", argno 1 // id: %1
  %2 = metatype $@thin Float.Type
  %3 = metatype $@thick Float.Type                // user: %13
  %4 = alloc_stack $Float                         // users: %5, %15, %13
  store %0 to %4 : $*Float                        // id: %5
  %6 = integer_literal $Builtin.IntLiteral, 0     // user: %9
  %7 = metatype $@thin Float.Type                 // user: %9
  // function_ref Float.init(_builtinIntegerLiteral:)
  %8 = function_ref @$sSf22_builtinIntegerLiteralSfBI_tcfC : $@convention(method) (Builtin.IntLiteral, @thin Float.Type) -> Float // user: %9
  %9 = apply %8(%6, %7) : $@convention(method) (Builtin.IntLiteral, @thin Float.Type) -> Float // user: %11
  %10 = alloc_stack $Float                        // users: %11, %14, %13
  store %9 to %10 : $*Float                       // id: %11
  // function_ref static FloatingPoint.> infix(_:_:)
  %12 = function_ref @$sSFsE1goiySbx_xtFZ : $@convention(method) <τ_0_0 where τ_0_0 : FloatingPoint> (@in_guaranteed τ_0_0, @in_guaranteed τ_0_0, @thick τ_0_0.Type) -> Bool // user: %13
  %13 = apply %12<Float>(%4, %10, %3) : $@convention(method) <τ_0_0 where τ_0_0 : FloatingPoint> (@in_guaranteed τ_0_0, @in_guaranteed τ_0_0, @thick τ_0_0.Type) -> Bool // user: %16
  dealloc_stack %10 : $*Float                     // id: %14
  dealloc_stack %4 : $*Float                      // id: %15
  %16 = struct_extract %13 : $Bool, #Bool._value  // user: %17
  cond_br %16, bb1, bb2                           // id: %17

bb1:                                              // Preds: bb0
  br bb3                                          // id: %18

bb2:                                              // Preds: bb0
  br bb3                                          // id: %19

bb3:                                              // Preds: bb1 bb2
  return %0 : $Float                              // id: %20
} // end sil function 'cond_empty'
```

Since `%0` is active in `bb0` and `bb3`, `%0_adj` must be somehow be propagated from `bb3_adj` to `bb0_adj` - perhaps via the adjoint value map for `bb1` and `bb2`.

Reply:
Resolved in https://github.com/apple/swift/pull/25057/commits/bd8d3190290840d0d05b6f32ee804072beab305f.
The solution is to track "dominated active values", not just "used active values" in original basic blocks.